pkg/auth/dependency/webapp/authenticate_provider_impl.go

```diff
@@ -267,3 +291,136 @@ func (p *AuthenticateProviderImpl) SetLoginID(r *http.Request) (err error) {
 
 	return
 }
+
+// For sso, we create state when user choose the provider
+// The state id will be passed into the sso state
+// When user complete authorization and provider call the callback
+// The callback will try to obtain the webapp state id from the sso state
+// Currently we use the state to store error only
+func (p *AuthenticateProviderImpl) ChooseIdentityProvider(w http.ResponseWriter, r *http.Request, oauthProvider OAuthProvider) (writeResponse func(err error), err error) {
+	s := p.makeState(r.URL.Query().Get("x_sid"))
+	var authURI string
+	writeResponse = func(err error) {
+		s.SetError(err)
+		if e := p.StateStore.Set(s); e != nil {
+			panic(e)
+		}
+		if err != nil {
+			RedirectToPathWithQuery(w, r, "/login", url.Values{
+				"x_sid": []string{s.ID},
+			})
+		} else {
+			http.Redirect(w, r, authURI, http.StatusFound)
+		}
+	}
+
+	if oauthProvider == nil {
+		panic("incorrect idp")
+	}
+
+	redirectURI := "/"
+	if u, err := getRedirectURI(r); err == nil {
+		redirectURI = u
+	}
+	nonce := sso.GenerateOpenIDConnectNonce()
+	hashedNonce := crypto.SHA256String(nonce)
+	state := sso.State{
+		Action: "login",
+		LoginState: sso.LoginState{
+			OnUserDuplicate: model.OnUserDuplicateAbort,
+		},
+		HashedNonce: hashedNonce,
+		OAuthAuthorizationCodeFlowState: sso.OAuthAuthorizationCodeFlowState{
+			CallbackURL: redirectURI,
+		},
+		WebAppStateID: s.ID,
+	}
+	encodedState, err := p.SSOProvider.EncodeState(state)
+	if err != nil {
+		return
+	}
+	authURI, err = oauthProvider.GetAuthURL(state, encodedState)
+	return
+}
+
+func (p *AuthenticateProviderImpl) HandleSSOCallback(w http.ResponseWriter, r *http.Request, oauthProvider OAuthProvider) (writeResponse func(error), err error) {
+	var callbackURL string
+	var sid string
+	writeResponse = func(err error) {
+		if err != nil {
+			// try to obtain state id from sso state
+			// create new state if failed
+			s := p.makeState(sid)
```

Comment:
I don't think we have sufficient information to re-create the state? Later on, we would have deeper integration with OAuth APIs, that the OAuth parameters may be lost if the state is expired (e.g. step up, redirect URI).